content/posts/2021/07-15-salt-nornir-intro.md

```diff
@@ -0,0 +1,229 @@
+
+Proxy Minions and Normal minions use execution modules to provide SALT Master and ultimately end-user with functionality to manage target systems/devices. For example, latest (at the time) version of SALT 3003.1 shipped with 529 execution modules each containing several functions. Many execution modules can be used by proxy-minions.
+
+## SALTSTACK how to use it
+
+For Human beings, preferred way of interacting with SALT is a collection of CLI utilities that you invoke on Master, main ones being `salt` and `salt-run`.
+
+For machines/scripts SALT exposes native [Python API](https://docs.saltproject.io/en/latest/ref/clients/index.html#client-apis) same as Nornir. In addition to that, SALT can run REST API server, which acts as a thin wrapper around Python API.
+
+Here is an example of running "clock" shell command on the remote Linux machine called *srv-1* through the minion using `salt` utility on SALT Master:
+
+```
+salt srv-1 cmd.run 'clock'
+```
+
+Or, example of running "show clock" command from SALT-Master on the remote network router called *router-1* managed over SSH by NAPALM Proxy Minion:
+
+```
+salt router-1 net.cli "show clock"
+```
+
+Or same example as above but using [Python Local Client API](https://docs.saltproject.io/en/latest/ref/clients/index.html#localclient) on SALT Master:
+
+```
+import salt.client
+
+client = salt.client.LocalClient()
+
+response = client.cmd(tgt="router-1", fun="net.cli", arg=["show clock"])
+```
+
+Another option to run SALT commands from CLI terminal on local machine could be [Pepper Library](https://github.com/saltstack/pepper), which leverages SALT REST API server.
+
+Worth noting that various SALTSTACK web GUI application were developed as well.
+
+## How Nornir fits the picture
+
+SALTSTACK gained prominent capabilities to manage network devices a while ago, mainly thanks to development of [NAPALM proxy-minion](https://docs.saltproject.io/en/latest/topics/network_automation/index.html#napalm) module in year 2016. 
+
+However, the main drawback of proxy-minion for network automation, or better say, the main drawback of three tier hierarchy for network devices automation is that each network device requires to run dedicated proxy-minion process, each consuming ~100Mbyte of RAM.
+
+If you have somewhat small network of say 50-100 devices (routers, switches, firewalls etc.) you might end up running 50-100 proxy minion processes (one for each managed devices) consuming about 4-10Gbyte of RAM combined. 
+
+If your network is of the bigger size and has about 500-1000 devices in it, you might need 40-100Gbyte of RAM to run your proxy-minions.
+
+Moving on to Service Provider or big Enterprise - network can grow up to several thousands of devices easily. In that case, resources to run proxy-minions would become substantial.
```

Comment:
I'd slightly rephrase:

"For a service provider or big enterprise with thousands of devices, the resources needed for the proxy-minions can be significant."